test/test_mixed_null_assertions.cpp

```diff
@@ -78,3 +78,224 @@ TEST(List_Mixed_do_insert)
     list.insert_null(0);
     list.insert(0, Mixed("hello world"));
 }
+
+TEST(Mixed_List_unresolved_as_null)
+{
+    Group g;
+    auto t = g.add_table("foo");
+    t->add_column_list(type_Mixed, "mixeds");
+    auto obj = t->create_object();
+    auto obj1 = t->create_object();
+
+    auto list = obj.get_list<Mixed>("mixeds");
+
+    list.insert_null(0);
+    list.insert(1, Mixed{"test"});
+    list.insert(2, obj1);
+    obj1.invalidate();
+
+    CHECK(list.size() == 3);
+
+    {
+        // find all mixed nulls or unresolved link should work the same way
+        std::vector<size_t> found;
+        auto check_results = [&](std::vector<size_t> expected) -> bool {
+            if (found.size() != expected.size())
+                return false;
+
+            std::sort(found.begin(), found.end());
+            std::sort(expected.begin(), expected.end());
+            for (size_t i = 0; i < found.size(); ++i) {
+                if (found[i] != expected[i])
+                    return false;
+            }
+            return true;
+        };
+        list.find_all(realm::null(), [&](size_t pos) {
+            found.push_back(pos);
+        });
+        CHECK(check_results({0, 2}));
+        found = {};
+        list.find_all(obj1, [&](size_t pos) {
+            found.push_back(pos);
+        });
+        CHECK(check_results({2}));
+    }
+
+    {
+        // find null or find unresolved link diverge, different objects should be returned
+        auto index = list.find_any(realm::null());
+        CHECK(index == 0);
+        index = list.find_first(obj1);
+        CHECK(index == 2);
+        // but both should look like nulls
+        CHECK(list.is_null(0));
+        CHECK(list.is_null(2));
+    }
+
+    {
+        std::vector<size_t> indices{0, 1, 2};
+        list.sort(indices);
+        CHECK(indices.size() == 3);
+        CHECK(indices.at(0) == 0);
+        CHECK(indices.at(1) == 2);
+        CHECK(indices.at(2) == 1);
+        CHECK(list.is_null(indices[0]));
+        CHECK(list.is_null(indices[1]));
+        CHECK(!list.is_null(indices[2]));
+    }
+
+    {
+        std::vector<size_t> indices{0, 1, 2};
+        list.distinct(indices);
+        CHECK(indices.size() == 2);
+        CHECK(indices.at(0) == 0);
+        CHECK(indices.at(1) == 1);
+        CHECK(list.is_null(indices[0]));
+        CHECK(!list.is_null(indices[1]));
+        CHECK(list.find_any(realm::null()) == 0);
+    }
+
+    {
+        list.remove(0);
+        CHECK(list.find_any(obj1) == 1);
+        list.remove(1);
+        CHECK(list.find_any(realm::null()) == npos);
+        CHECK(list.size() == 1);
+    }
+}
+
+TEST(Mixed_Set_unresolved_links)
```

Comment:
Not sure we want to have a test that checks the broken behaviour, but I guess it is ok to leave it in.

Reply:
I prefer to keep it, I think we should think about building a set of expired links ... but that is for a second PR. Or alternatively, I don't know how expensive is with flx sync to refresh the object. Surely we can change the file format for a small corner case like this.